File: Day3/Python/rucksack.py
from string import ascii_lowercase as ascl
import numpy as np

class Ruck:
  def __init__(self, filename: str) -> None:
    self.filename = filename
    self.priority = {}
    for i, letter in enumerate(ascl + ascl.upper()):
      self.priority[letter] = i + 1

    self.rucksacks = []
    with open(self.filename, 'r') as f:
      for l in f:
        self.rucksacks.append(l)

  def part1(self) -> int:
 
    # Find duplicates
    dups = []
    for sack in self.rucksacks:
      first_half = sack[:int(len(sack) // 2)]
      second_half = sack[int(len(sack) // 2):]
  
      for i in first_half:
        if i in second_half:
          dups.append(i)
          break
  
    # Find sum
    sum = 0
    for c in dups:
      sum += self.priority[c]
      
    return sum
  
  def part2(self) -> int: 
    # Find badges
    badges = []
    for i in np.arange(0, len(self.rucksacks), 3):
      for c in self.rucksacks[i]:
        if c in self.rucksacks[i+1] and c in self.rucksacks[i+2]:
          badges.append(c)
          break

    # Find sum
    sum = 0
    for c in badges:
      sum += self.priority[c]

    return sum
    

File: Day3/Julia/rucksack.jl
module Rucksack

    priority = Dict{Char, Int}()
    for (value, letter) in enumerate(vcat( collect('a':'z'), collect('A':'Z')))
        priority[letter] = value
    end

    # function to get lines from a file
    function get_file( filename::String )::Vector{String}
        file = open(filename)
        lines = Vector{String}()

        for number in readlines(file)
            push!(lines, number)
        end 
        close(file)
        return lines
    end


    function part_1( rucksack::Vector{String} )::Int
        duplicates = Vector{Char}()
        sum::Int = 0

        for item_set in rucksack
            first_half = item_set[ 1:Int(length(item_set)/2) ]
            second_half = item_set[ Int(length(item_set)/2)+1:end ]

            for item in first_half
                if occursin(item, second_half)
                    push!(duplicates, item)
                    break
                end
            end
        end

        for item in duplicates
            sum += priority[item]
        end

        return sum
    end

    function part_2( rucksack::Vector{String} )::Int
        badges = Vector{Char}()
        sum::Int = 0

        for idx in 1:3:length(rucksack)
            for item in rucksack[idx]
                if occursin(item, rucksack[idx+1]) && occursin(item, rucksack[idx+2])
                    push!(badges, item)
                    break
                end
            end
        end

        for item in badges
            sum += priority[item]
        end

        return sum
    end

end